Data/Aeson/Types/FromJSON.hs

```diff
@@ -178,18 +178,46 @@ parseRealFloat _        Null       = pure (0/0)
 parseRealFloat expected v          = typeMismatch expected v
 {-# INLINE parseRealFloat #-}
 
+truncateIntegral :: Integral a => String -> Scientific -> Parser a
+truncateIntegral expected s
+    | Scientific.isInteger s = pure (truncate s)
+    | otherwise               =
+        fail $ "Floating number specified for " ++ expected ++ ": " ++ show s
+{-# INLINE truncateIntegral #-}
+
 parseIntegral :: Integral a => String -> Value -> Parser a
-parseIntegral expected = withScientific expected $ pure . truncate
+parseIntegral expected = withScientific expected $ truncateIntegral expected
 {-# INLINE parseIntegral #-}
 
+truncateBoundedIntegral :: Integral a => String -> Scientific -> Parser a
+truncateBoundedIntegral expected s
+    | Scientific.isInteger s =
+        let x = truncate s
```

Comment:
Why use `truncate` here? If we've already determined `s` is integer truncating will not have an effect anymore.

I would recommend using [toBoundedIntege](http://hackage.haskell.org/package/scientific-0.3.4.8/docs/Data-Scientific.html#v:toBoundedInteger) here.